index.rst

```diff
@@ -51,11 +51,182 @@
+Implementing a client/server ``Butler.put()`` is difficult because there is no way to convert the Python in-memory dataset to serialized form without involving the formatter infrastructure that is called within ``Datastore``.
+One option is for the server implementation of ``Butler.put()`` to not take the dataset at all, but to instead store a placeholder entry in the datastore registry and then return a signed URL, along with the ``DatasetRef``, that can be used by the client code to push the file.
+The client code would then use a local Datastore implementation to create the file from the relevant formatter and then upload it.
+The downside of this approach is that it is not immediately clear how to handle composite disassembly (where the file is split into multiple components and each is stored separately in the file store) since that is a datastore configuration.
+
+A ``Butler.get()`` would most logically be handled as a call to ``Butler.getURIs()`` (probably via the datastore ``getStoredItemsInfo`` method) to obtain the signed URL, download the file, and then use a local Datastore implementation to read that file.
+This could conceivably leverage the purported "caching datastore" concept.
+
+Even if all this is made to work, users still expect to be able to use many of the registry methods for querying the collections, datasets and dimension records.
+This suggests that it might be better to implement the client/server registry first and build on that, and then subsequently add explicit ``Butler`` overrides if performance is an issue (for example if ``Butler`` calls registry methods in a loop there may be significant run time overheads).
+
+Manager Classes
+---------------
+
+The final option is to implement each of the manager classes as client/server implementations.
+The registry is already designed to support pluggable managers and they are already designed to isolate database access.
+This seems like the cleanest way forward but the interfaces are very low level and this makes it significantly harder for non-Butler clients to do anything useful with the data being returned.
+It also raises the possibility of the interfaces being very slow when called in loops and may require significant caching of results and also the addition of new methods that move loops into the server.
+
+
+Access Controls
+===============
+
+Part of the benefit of using a client/server approach is that the server can control access to collections and datastores without having to use fine-grained database permissions on specific tables or add ACLs to the object store.
+
+This does though mean that there must be code in the server that can take the user name and determine which information can be used.
+This does not simply mean checking that the collection name includes the user name since the checks must also be able to look at collections that have group access controls (one person may wish to give access of their processed data to another user but no-one else).
```

Comment:
I do think one of the next big steps here is to figure out the ownership model for butler entities:
 - what kinds of butler objects (datasets, dataset types, collections, dimension records, ...) can be have ownership or other privileges associated with them;
 - what those privileges are for each kind of entity (e.g. read, insert, delete, update, various kinds of grant-privileges, various kinds reference-this-entity-in-my-entity)
 - what kinds of roles (users, groups, lists of users, lists of groups) can have those privileges
 - where we store that mapping (presumably in new database tables and columns)
 - how to retrieve privilege information efficiently at the points it needs to be applied.

If our goal for the data previews is a client/server butler where everybody can read everything and nobody can write anything, that's a _lot_ simpler to stand up first, and then I'd have no problem punting these questions to the future.  A lot of other problems (e.g. `butler.put` transactions) also go away if we're talking about a read-only butler, but I think a lot of those are important questions to wrestle with now in order to get the architecture something we can extend down the line.  I'm worried some of these questions fit that characterization, too, but not _as_ worried.